api/include/opentelemetry/metrics/meter.h

```diff
@@ -0,0 +1,323 @@
+#pragma once
+
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/variant.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+#include "opentelemetry/version.h"
+
+#include "opentelemetry/metrics/async_instruments.h"
+#include "opentelemetry/metrics/instrument.h"
+#include "opentelemetry/metrics/sync_instruments.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace metrics
+{
+/**
+ * Handles instrument creation and provides a facility for batch recording.
+ *
+ * This class provides methods to create new metric instruments, record a
+ * batch of values to a specified set of instruments, and collect
+ * measurements from all instruments.
+ *
+ */
+class Meter
+{
+public:
+  virtual ~Meter() = default;
+
+  /**
+   * Creates, adds to private metrics container, and returns a DoubleCounter with "name."
+   *
+   * @param name the name of the new DoubleCounter.
+   * @param description a brief description of what the DoubleCounter is used for.
+   * @param unit the unit of metric values following https://unitsofmeasure.org/ucum.html.
+   * @param enabled a boolean value that turns on or off the metric instrument.
+   * @return a shared pointer to the created DoubleCounter.
+   * @throws NullPointerException if {@code name} is null
+   * @throws IllegalArgumentException if a different metric by the same name exists in this meter.
+   * @throws IllegalArgumentException if the {@code name} does not match spec requirements.
+   */
+  virtual opentelemetry::nostd::shared_ptr<SynchronousInstrument> NewDoubleCounter(
+      nostd::string_view name,
+      nostd::string_view description,
+      nostd::string_view unit,
+      const bool enabled)
```

Comment:
Shouldn't this be pure virtual? This will keep the API layer as slim as possible.

For traces we followed this pattern:
1. Abstract (interface) classes with pure virtual methods.
2. Noop implementations.

It's a bit confusing to have default implementations of those virtual methods, I'm not sure what this means for SDK implementors and how they are supposed to hook in.

Reply:
I can make the functions purely virtual if desired. I asked @reyang about this a while back and I was told to err on the side of virtual functions with no-op implementations. Maybe I misunderstood this. Would others like for these to be purely virtual functions?

Reply:
I had similar questions in [other PRs](https://github.com/open-telemetry/opentelemetry-cpp/pull/161#discussion_r454625940).

For interfaces that are not suppose to be instantiated, we prefer pure virtual.

Reply:
Specifically for this case, making it pure virtual would enforce the SDK to provide an implementation, which I think is a good thing to have.